activerecord/lib/active_record/autosave_association.rb

```diff
@@ -273,7 +282,23 @@ def changed_for_autosave?
       new_record? || has_changes_to_save? || marked_for_destruction? || nested_records_changed_for_autosave?
     end
 
+    # Returns whether or not this record is already being saved outside of the
+    # current autosave callback
+    def already_saving?
+      @saving
+    end
+
     private
+      # Track if this record is already being saved. If it is being saved we
+      # can skip saving it in the autosave callbacks
+      def with_saving_state_tracking
```

Comment:
Little bikeshedding: any thoughts on naming this method `saving` and the corresponding predicate method `saving?`?

```ruby
def saving?
  @saving
end

private
  def saving
    previously_saving, @saving = @saving, true
    yield
  ensure
    @saving = previously_saving
  end
```

Reply:
I think the `saving?` method is better than `already_saving?`.
It could be called outside of the autosave code where `already_saving?` might be confusing.
I like that `saving` is named similarly to `saving?` because it makes it more clear they belong together.
Although I'm afraid they might conflict with existing code bases that already have these methods defined.